### PR TITLE
XMC4xxx pinctrl driver

### PR DESCRIPTION
**Pull request by talih0**
Adds pinctrl driver and modifies uart to use driver.

### CODE REVIEW

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ * Copyright (c) 2022 Schlumberger
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/pinctrl/xmc4xxx-pinctrl.h>
+
+#include <devicetree.h>
+#include <xmc_gpio.h>
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:11]:
```suggestion
#include <zephyr/devicetree.h>
#include <zephyr/drivers/pinctrl.h>
#include <zephyr/dt-bindings/pinctrl/xmc4xxx-pinctrl.h>

#include <xmc_gpio.h>
```

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio3)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio4)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio5)),
+	(void *)REG_OR_NULL(DT_NODELABEL(gpio6)),
+};
+
+static int pinctrl_configure_pin(const pinctrl_soc_pin_t *pinctrl)
+{
+	int port_id, pin, alt_fun, hwctrl;
+	uint32_t pinmux = pinctrl[0];
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:28]:
```suggestion
static int pinctrl_configure_pin(pinctrl_soc_pin_t pinmux)
{
	int port_id, pin, alt_fun, hwctrl;
```

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	int ret = 0;
+
+	for (int i = 0; i < pin_cnt; i++) {
+		ret = pinctrl_configure_pin(pins++);
+		if (ret < 0) {
+			break;
+		}
+	}
+
+	return ret;
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:96]:
```suggestion
	for (uint8_t i = 0U; i < pin_cnt; i++) {
		int ret = pinctrl_configure_pin(*pins++);
		if (ret < 0) {
			break;
		}
	}

	return 0;
```

File: drivers/serial/uart_xmc4xxx.c [hunk 4]
```diff
@@ -66,12 +70,16 @@ static const struct uart_driver_api uart_xmc4xxx_driver_api = {
...
+PINCTRL_DT_INST_DEFINE(index);						\
 static struct uart_xmc4xxx_data xmc4xxx_data_##index = {		\
 	.config.baudrate = DT_INST_PROP(index, current_speed)		\
 };									\
 									\
 static const struct uart_xmc4xx_config xmc4xxx_config_##index = {	\
 	.uart = (XMC_USIC_CH_t *)DT_INST_REG_ADDR(index),		\
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(index),			\
+	.input_src = XMC4XXX_PINMUX_GET_ALT(				\
+		DT_INST_PROP_BY_PHANDLE_IDX(index, rx_pin, 0, pinmux)), \
```
Comment (gmarull) [inline drivers/serial/uart_xmc4xxx.c:82]: why is this required?

questions/suggestions:
- could `XMC_UART_CH_SetInputSource` be handled by pinctrl? (you can store periph reg to perform per peripheral actions)
- If you need pin info (e.g. as it happens for i2c bus recovery in some drivers), use `xxx-gpios` and keep it independent of pinctrl

File: dts/arm/infineon/xmc4500.dtsi [hunk 5]
```diff
@@ -23,3 +24,152 @@
...
+		drive-strength = "strong-soft-edge";
+		drive-push-pull;
+		output-high;
+	};
+	uart_tx_p1_5_u0c0: uart_tx_p1_5_u0c0 {
+		pinmux = <XMC4XXX_PINMUX_SET(1, 5, 2)>;
+		drive-strength = "strong-soft-edge";
+		drive-push-pull;
+		output-high;
+	};
```
Comment (gmarull) [inline dts/arm/infineon/xmc4500.dtsi:86]: - predefined nodes should be in a separate header ideally, e.g. `xmc4500-pinctrl.dtsi`. Also, options are tipically package dependent (not SoC only). I see 100/144 pin packages for this SoC.
- Defaults should be avoided in general, except there's one single valid combination. Please justify.

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# Copyright (c) 2022, Andriy Gelman <andriy.gelman@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Infineon XMC4XXX Pinctrl container node
```
Comment (gmarull) [inline dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml:5]: please extend description, see other vendors

File: soc/arm/infineon_xmc/4xxx/pinctrl_soc.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_SOC_ARM_INFINEON_XMC_4XXX_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_ARM_INFINEON_XMC_4XXX_PINCTRL_SOC_H_
+
+#include <zephyr/dt-bindings/pinctrl/xmc4xxx-pinctrl.h>
+#include <zephyr/types.h>
+
+#include <devicetree.h>
```
Comment (gmarull) [inline soc/arm/infineon_xmc/4xxx/pinctrl_soc.h:13]:
```suggestion
#include <stdint.h>

#include <zephyr/devicetree.h>
#include <zephyr/dt-bindings/pinctrl/xmc4xxx-pinctrl.h>
```

**Review by gmarull — CHANGES_REQUESTED** [review]
Thanks for this contribution, please check my comments

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 3 above)
Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:96]: will update. Why do you prefer to return 0, and not forward the error?

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml (hunk 6 above)
Comment (talih0) [reply dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml:5]: will update.

File: dts/arm/infineon/xmc4500.dtsi (hunk 5 above)
Comment (talih0) [reply dts/arm/infineon/xmc4500.dtsi:86]: > * predefined nodes should be in a separate header ideally, e.g. `xmc4500-pinctrl.dtsi`. Also, options are tipically package dependent (not SoC only). I see 100/144 pin packages for this SoC.

I'll put the pre-defined nodes in a separate file and use a name with a pin package. Btw, I also have a pull request to rename xmc4500.dtsi:
https://github.com/zephyrproject-rtos/zephyr/pull/47161/commits/e7c0d85526166b7716a51e4d640f87fd344e731d

Comment (talih0) [reply dts/arm/infineon/xmc4500.dtsi:86]: > * Defaults should be avoided in general, except there's one single valid combination. Please justify.

I'll move the pin config settings to the board dts.

File: drivers/serial/uart_xmc4xxx.c (hunk 4 above)
Comment (talih0) [reply drivers/serial/uart_xmc4xxx.c:82]: > why is this required?

The peripherals on the xmc4500 also use a mux to connect to different gpio pins. I'll think about the two ways you proposed.

> questions/suggestions:
> 
>     * could `XMC_UART_CH_SetInputSource` be handled by pinctrl? (you can store periph reg to perform per peripheral actions)

Yes, but it didn't quite feel right to have peripheral specific code in the pinctrl driver. It would also need to handle different types of peripherals (USIC, CAN, Ethernet). Are there any other pinctrl drivers that do this?

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 2 above)
Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:28]: yes, that's much cleaner.

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 3 above)
Comment (gmarull) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:96]: updated example, missed one return

File: drivers/serial/uart_xmc4xxx.c (hunk 4 above)
Comment (talih0) [reply drivers/serial/uart_xmc4xxx.c]: > * If you need pin info (e.g. as it happens for i2c bus recovery in some drivers), use `xxx-gpios` and keep it independent of pinctrl

I added a property input-src in a separate commit. Let me know if this is ok.

**Comment by talih0** [comment]
I added the gpio nodes in the device tree for fetching the register addresses. But there are no xmc gpio drivers for zephyr yet.
I'll make a separate pull request for the gpio drivers; it may be better to merge gpio first before pinctrl.

**Comment by talih0** [comment]
I'll resend after the gpio drivers is merged:
https://github.com/zephyrproject-rtos/zephyr/pull/47905

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+        invert-input:
+          description: |
+            Inverts the input.
+          required: false
+          type: boolean
+
+        hwctrl:
+          description:
+            Pre-assigns hardware control of the pin to a certain peripheral.
```
Comment (mbolivar-nordic) [inline dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml:110]: https://docs.zephyrproject.org/latest/build/dts/bindings.html#general-rules says:

If your description is short, it’s fine to use this style:

description: my short string

If your description is long or spans multiple lines, you must use this style:

description: |
  My very long string
  goes here.
  Look at all these lines!

This | style prevents YAML parsers from removing the newlines in multi-line descriptions. This in turn makes these long strings display properly in the [Bindings index](https://docs.zephyrproject.org/latest/build/dts/api/bindings.html#devicetree-binding-index).

Do not use any other style for long or multi-line strings.

**Review by mbolivar-nordic — DISMISSED** [review]
string style nit

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml (hunk 8 above)
Comment (talih0) [reply dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml]: thanks, I've fixed it in the latest push.

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 9]
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2022 Schlumberger
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gpio/gpio_xmc4xxx.h>
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:7]: why is this needed? please keep gpio/pinctrl independent

File: drivers/pinctrl/Kconfig.xmc4xxx [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 Schlumberger
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_XMC4XXX_PINCTRL := infineon,xmc4xxx-pinctrl
```
Comment (gmarull) [inline drivers/pinctrl/Kconfig.xmc4xxx:4]: can be removed

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+	int port_id, pin, alt_fun, hwctrl;
+	XMC_GPIO_CONFIG_t pin_cfg = {0};
+
+	port_id = XMC4XXX_PINMUX_GET_PORT(pinmux);
+	if (port_id >= ARRAY_SIZE(gpio_ports)) {
+		return -EINVAL;
+	}
+
+	gpio_device = gpio_ports[port_id];
+	if (gpio_device == NULL || !device_is_ready(gpio_device)) {
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:36]: device_is_ready handles NULL input

File: dts/arm/infineon/xmc4500_F100x1024-pinctrl.dtsi [hunk 12]
```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2022 Schlumberger
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* File autogenerated with python script */
```
Comment (gmarull) [inline dts/arm/infineon/xmc4500_F100x1024-pinctrl.dtsi:6]: which python script?

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        invert-input:
+          description: Inverts the input.
+          required: false
+          type: boolean
+
+        hwctrl:
+          description: Pre-assigns hardware control of the pin to a certain peripheral.
+          required: false
+          type: string
+          default: "disabled"
```
Comment (gmarull) [inline dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml:110]: defaults need to be justified.

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+          required: true
+          type: int
+
+        drive-strength:
+          description: |
+            Drive strength of the output pin. Following options as in XMC_GPIO_OUTPUT_STRENGTH
+            See xmc4_gpio.h. This only has an effect if the pin is in drive-push-pull mode.
+          required: false
+          type: string
+          default: "strong-medium-edge"
```
Comment (gmarull) [inline dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml:90]: ditto

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 9 above)
Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:7]: > why is this needed? please keep gpio/pinctrl independent

I was following the example in pinctrl_stm32.c.

I used the gpio data structure to get the gpio port base address.
But I can also compute it from the pinctrl base address. Do you prefer this approach?

File: drivers/pinctrl/Kconfig.xmc4xxx (hunk 10 above)
Comment (talih0) [reply drivers/pinctrl/Kconfig.xmc4xxx:4]: yep, will update.

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 11 above)
Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:36]: thanks, will update.
Do I even need to check that the gpio device has been initialized?
For example, this is checked in pinctrl_smt32.c, but not done in pinctrl_b91.c

File: dts/arm/infineon/xmc4500_F100x1024-pinctrl.dtsi (hunk 12 above)
Comment (talih0) [reply dts/arm/infineon/xmc4500_F100x1024-pinctrl.dtsi:6]: > which python script?

it's my own script that I used to parse Infineon headers;
https://pastebin.com/6ED3GKfE

shall I just remove this comment altogether?

File: dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml (hunk 13 above)
Comment (talih0) [reply dts/bindings/pinctrl/infineon,xmc4xxx-pinctrl.yaml:110]: I'll remove the defaults.

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 9 above)
Comment (gmarull) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:7]: that's the only case it's this way, mainly for historical reasons and it should be fixed.

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 11 above)
Comment (gmarull) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:36]: Yes, otherwise it's UB.

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	XMC_GPIO_PORT_t *gpio_port;
+
+	port_id = XMC4XXX_PINMUX_GET_PORT(pinmux);
+	if (port_id >= ARRAY_SIZE(gpio_devs)) {
+		return -EINVAL;
+	}
+
+	if (!device_is_ready(gpio_devs[port_id])) {
+		return -ENODEV;
+	}
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:37]: why is this required?

File: drivers/pinctrl/pinctrl_xmc4xxx.c [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+#include <xmc_gpio.h>
+
+static const struct device *const gpio_devs[] = {
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio0)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio1)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio2)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio3)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio4)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio5)),
+};
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xmc4xxx.c:22]: this array is only used to check port_id size?

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 15 above)
Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:37]: I use it as a check to make sure gpioX has been defined in the device tree.

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 16 above)
Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:22]: It's used in the device_is_ready() above.

File: drivers/pinctrl/pinctrl_xmc4xxx.c (hunk 15 above)
Comment (gmarull) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:37]: but why? I still don't get the value of adding this.

Comment (talih0) [reply drivers/pinctrl/pinctrl_xmc4xxx.c:37]: I thought you said to keep it in: 
https://github.com/zephyrproject-rtos/zephyr/pull/47581#discussion_r949131795

Probably a misunderstanding on my part. I'll remove it in the next push

File: soc/arm/infineon_xmc/CMakeLists.txt [hunk 17]
```diff
@@ -3,4 +3,5 @@
 # Copyright (c) 2020 Linumiz
 # Author: Parthiban Nallathambi <parthiban@linumiz.com>
 
+zephyr_include_directories(${ZEPHYR_BASE}/drivers)
```
Comment (gmarull) [inline soc/arm/infineon_xmc/CMakeLists.txt:6]: why is this needed?

File: drivers/serial/Kconfig.xmc4xxx [hunk 18]
```diff
@@ -8,5 +8,6 @@ config UART_XMC4XXX
 	default y
 	depends on DT_HAS_INFINEON_XMC4XXX_UART_ENABLED
 	select SERIAL_HAS_DRIVER
+	select GPIO
```
Comment (gmarull) [inline drivers/serial/Kconfig.xmc4xxx:11]: why is this needed? I can't see driver changes using the GPIO API

File: boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts [hunk 19]
```diff
@@ -45,10 +46,16 @@
...
 	current-speed = <115200>;
+	pinctrl-0 = <&uart_tx_p0_1_u1c1 &uart_rx_p0_0_u1c1>;
+	pinctrl-names = "default";
 	input-src = "DX0D";
 	status = "okay";
 };
 
+&gpio0 {
+	status = "okay";
+};
```
Comment (gmarull) [inline boards/arm/xmc45_relax_kit/xmc45_relax_kit.dts:57]: ditto

File: drivers/serial/Kconfig.xmc4xxx (hunk 18 above)
Comment (talih0) [reply drivers/serial/Kconfig.xmc4xxx]: yep, it's no longer needed after I removed gpio dependency, thanks. (same for the other points in your review).

**Review by gmarull — DISMISSED** [review]
thanks for the continued effort

**Comment by talih0** [comment]
> thanks for the continued effort

and thank you for reviewing!

**Comment by talih0** [comment]
ping, is it possible to push this?
I'd like to rebase PR https://github.com/zephyrproject-rtos/zephyr/pull/48540 on top this.

**Comment by gmarull** [comment]
@talih0 please rebase there seem to be conflicts

**Review by gmarull — APPROVED** [review]

**Comment by gmarull** [comment]
@carlescufi ping

**Review by nashif — APPROVED** [review]
